### PR TITLE
WIP Indexing autograd and tests, new loss functions

### PR DESCRIPTION
**Pull request by syurkevi**
Adds new indexing functions to address #43.  Theses are needed to implement Losses ans SVM from #34 and #6 those changes are next in line.
Operator overloading currently avoided in favor of new __select_index ()__ and __set_index functions()__. Will change to preferred solution as part of review.
Also brings in google test framework for testing. examples/autograd.cpp has been converted to (non-comprehensive) tests.

### CODE REVIEW

File: include/af/autograd/Functions.hpp [hunk 1]
```diff
@@ -44,6 +44,8 @@ namespace af {
         Variable operator <=(const Variable &lhs, const double &rhs);
 
         Variable operator !(const Variable &input);
+        Variable select_index(const Variable &input, const Variable &idx);
+        Variable set_index(const Variable &input, const Variable &idx, const Variable &vals);
```
Comment (pavanky) [inline include/af/autograd/Functions.hpp]: I think `lookup` and `assign` are better. Or  go with`gather` vs `scatter` like tensorflow.

File: src/autograd/Functions.cpp [hunk 2]
```diff
@@ -117,6 +117,35 @@ namespace af {
...
 
+        Variable select_index(const Variable &input, const Variable &idx)
+        {
+            af::array result = input.array()(idx.array());
+            af::array mask   = af::constant(0, input.dims());
+            mask(idx.array()) = 1;
+
+            auto grad_func = [](std::vector<Variable> &inputs, const Variable &grad_output) {
+                auto grad = inputs[2].array();
+                auto grad_mask = af::where(grad);
```
Comment (pavanky) [inline src/autograd/Functions.cpp]: Why are you converting `idx` into a mask and converting it back into an index again? This is making unnecessary copies.

File: src/autograd/Functions.cpp [hunk 3]
```diff
@@ -117,6 +117,35 @@ namespace af {
...
+                grad(grad_mask) *= grad_output.array();
+
+                inputs[0].addGrad(Variable(grad, false));
+            };
+            return Variable(result, {input, idx, Variable(mask, false)}, grad_func);
```
Comment (pavanky) [inline src/autograd/Functions.cpp]: You only need `idx` or `mask`. Not both.

File: src/autograd/Functions.cpp [hunk 4]
```diff
@@ -117,6 +117,35 @@ namespace af {
...
+        {
+            af::array result = input.array();
+            result(idx.array()) = vals.array();
+            af::array mask   = af::constant(1, input.dims(), s32);
+            mask(idx.array()) = 0;
+
+            auto grad_func = [](std::vector<Variable> &inputs, const Variable &grad_output) {
+                inputs[0].addGrad(inputs[3] * grad_output);
+            };
+            return Variable(result, {input, idx, vals, Variable(mask, false)}, grad_func);
```
Comment (pavanky) [inline src/autograd/Functions.cpp]: It is better to use just index and do the masking inside the grad_func. No reason to do the masking op until it is necessary.

**Review by pavanky — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 5]
```diff
@@ -35,3 +35,9 @@ set_target_properties(afml
   CXX_STANDARD 11)
 
 add_subdirectory(examples)
+
+option(PACKAGE_TESTS "Build tests" ON)
+if(PACKAGE_TESTS)
+    enable_testing()
+    add_subdirectory(tests)
+endif()
```
Comment (9prady9) [inline CMakeLists.txt:43]: You can avoid creating a separate option `PACKAGE_TESTS`. gtest automatically addes `BUILD_TESTING` which you can use.
